### DISCUSSION TITLE
open pdf warning

### DISCUSSION THREAD

**Discussion by zyc130130** [pymupdf/PyMuPDF#1153] [Q&A] [answered]
Hi,
  I try to open a pdf and get it's table of contents, but screen outputs some infomation "mupdf: invalid key in dict", what happen?
  By the way, method of "get_toc" it's result is right.

  use code:
      with open(file_name, "rb") as fr:
        file_name = fr.read()
      doc = fitz.open(stream=file_name, filetype="pdf")
      print(doc.get_toc())

screen output:
   mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict
mupdf: invalid key in dict

The pdf has attached.
[876912f8d1749473cefb41681c51fdbf.pdf](https://github.com/pymupdf/PyMuPDF/files/6829152/876912f8d1749473cefb41681c51fdbf.pdf)

Except your replay.

**Comment by JorjMcKie** [comment]
Very simple answer: the PDF is damaged in various places.
Try to repair it, e.g. via ´mutool clean -gggz file.pdf`, or simply live with the error messages.

**Comment by zyc130130** [comment]
> Very simple answer: the PDF is damaged in various places.
> Try to repair it, e.g. via ´mutool clean -gggz file.pdf`, or simply live with the error messages.

Thanks for your replay.
The pdf is broken, which sounds sad.
Now I try to get table of contents and index which in some pdfs, I used some method like "doc.get_toc", I know there is not bookmark in every pdf, even bookmark is not exactly equivalent to table of contents, so I use image classification technology to process images and locate the table of contents and index location. Do you have any suggestions for this requirement or a better way.

**Reply by JorjMcKie** [reply]
I don't understand: Bookmark that is not part of a table of contents? What is that?
And how can you locate a TOC via "image classification technology" that is not accessible via the PDF itself?
Are we talking about the same things at all?
* maybe you want to **_create_** a TOC?
* or insert links?

**Comment by zyc130130** [comment]
Yes, I want to creat TOC via some page named "contents".


I want to get some pages which are named "contents" or "index" and parse their text. I want to get contents-text-tree-structure which like bookmark(get via doc.get_toc). Using&nbsp;image classification to get which page num is page named "contents" or "index".


I means some bookmark is part of page "contents", example, bookmark is only have level 1, but pdf "contents" have level 2 and level 3. I know bookmark is created by somebody, he maybe don't write level 2.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "pymupdf/PyMuPDF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月19日(星期一) 下午4:02
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [pymupdf/PyMuPDF] open pdf warning (#1153)





 
I don't understand: Bookmark that is not part of a table of contents? What is that?
 And how can you locate a TOC via "image classification technology" that is not accessible via the PDF itself?
 Are we talking about the same things at all?
  
maybe you want to create a TOC?
 
or insert links?

**Reply by JorjMcKie** [reply]
Ok. This is not an easy task, because you must extract and interpret text.
And a lot of detail problems may occur here: "Contents" pages may be more than one, you have to ignore header lines and footer lines, etc.
If you in addition want to add more detailed bookmarks (like level 2+), things are getting even more complicated.
It will probably be better to scan the text of all document pages and try to find text headers yourself ... in many cases they have a different font or are bold, etc. or are prefixed with paragraph and section numbers like **_"3.4.2 Section ..."_**, and so on.
Good luck!

**Comment by zyc130130** [comment, marked as answer]
Yes, as you said, this is a complicated and hard task. There are many condition need to consider, like number of columns and so on.
Through image classification and pumupdf, I can extract some table of contents, I only extract part of the all PDF files due to the diversity of pdf table of contents, and there are some details problem which made it inaccurate.
I have try to use ocr, it is helped to improve the extraction range of all PDF files, but it was too slow, and some problems such as incorrect text might also be generated.




Thanks for you replay again.
Best wish to you!






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "pymupdf/PyMuPDF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年7月19日(星期一) 晚上6:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [pymupdf/PyMuPDF] open pdf warning (#1153)





 
Ok. This is not an easy task, because you must extract and interpret text.
 And a lot of detail problems may occur here: "Contents" pages may be more than one, you have to ignore header lines and footer lines, etc.
 If you in addition want to add more detailed bookmarks (like level 2+), things are getting even more complicated.
 It will probably be better to scan the text of all document pages and try to find text headers yourself ... in many cases they have a different font or are bold, etc. or are prefixed with paragraph and section numbers like "3.4.2 Section ...", and so on.
 Good luck!
